src/lib/crypto/kimchi_backend/gadgets/keccak.ml

```diff
@@ -0,0 +1,720 @@
+                in
+                let two_pow_8 = Bignum_bigint.(pow (of_int 2) (of_int 8)) in
+                let byte =
+                  Bignum_bigint.(
+                    (word - (word % offset_lo)) / offset_lo % two_pow_8)
+                in
+                Common.bignum_bigint_to_field (module Circuit) byte )
+          in
+          bytestring.(index) <- byte
+        done ;
+
+        (* TODO: Does this generate automatic generic gates to check decomposition? *)
+
+        (* Create a list containing the elements of bytestring from base_index to base_index + 7 *)
+        let word_bytes =
+          Array.to_list
+          @@ Array.sub bytestring ~pos:base_index ~len:bytes_per_word
+        in
+
+        (* Assert correct decomposition of bytes from state *)
+        check_bytes_to_word (module Circuit) state.(x).(y) word_bytes
+      done
+    done ;
+    Array.to_list bytestring
+
+  let xor (type f)
+      (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+      (input1 : Circuit.Field.t matrix) (input2 : Circuit.Field.t matrix) :
+      Circuit.Field.t matrix =
+    let open Circuit in
+    assert (
+      Array.length input1 = keccak_dim && Array.length input1.(0) = keccak_dim ) ;
+    assert (
+      Array.length input2 = keccak_dim && Array.length input2.(0) = keccak_dim ) ;
+    let output =
+      Array.make_matrix ~dimx:keccak_dim ~dimy:keccak_dim Field.zero
+    in
+    for y = 0 to keccak_dim - 1 do
+      for x = 0 to keccak_dim - 1 do
+        output.(x).(y) <-
+          Bitwise.bxor64 (module Circuit) input1.(x).(y) input2.(x).(y)
+      done
+    done ;
+    output
+
+  let _print (type f)
+      (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+      (state : Circuit.Field.t matrix) =
+    let open Circuit in
+    as_prover (fun () ->
+        for x = 0 to keccak_dim - 1 do
+          for y = 0 to keccak_dim - 1 do
+            let _elem =
+              Common.cvar_field_to_bignum_bigint_as_prover
+                (module Circuit)
+                state.(x).(y)
+            in
+            ()
+          done
+        done )
+end
+
+(* KECCAK HASH FUNCTION IMPLEMENTATION *)
+
+(* Performs the modulo operation, not remainder as in mod which instead computes integer remainder *)
+let modulo (x : int) (m : int) : int = ((x mod m) + m) mod m
+
+(* Pads a message M as:
+   * M || pad[x](|M|)
+   * Padding rule 0x06 ..0*..1.
+   * The padded message vector will start with the message vector
+   * followed by the 0*1 rule to fulfil a length that is a multiple of rate (in bytes)
+*)
+let pad_nist (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (message : Circuit.Field.t list) (rate : int) : Circuit.Field.t list =
+  let open Circuit in
+  (* Find out desired length of the padding in bytes *)
+  (* If message is already rate bits, need to pad full rate again *)
+  let extra_bytes = (rate / 8) - (List.length message mod rate) in
```

Comment:
This can be negative if e.g. `List.length messages = (8*k+1)*rate+1`

Reply:
Okay the longest test I had had 455 bits. Now I have one with more than 1088 bits to force it not only use more blocks but also break the typo-ed formula when padding. The new test uses 30k rows instead (2 blocks not only 1). Fixed now, both in nist and 10*1 (they were a copy paste of each other).